box-content-sdk/src/main/java/com/box/androidsdk/content/requests/BoxRequestDownload.java

```diff
@@ -203,60 +216,81 @@ public BoxDownload onResponse(Class clazz, BoxHttpResponse response) throws Ille
             String contentType = response.getContentType();
             long contentLength = -1;
 
-            if (response.getResponseCode() == HttpStatus.SC_OK || response.getResponseCode() == HttpStatus.SC_PARTIAL_CONTENT){
+            if (response.getResponseCode() == HttpStatus.SC_ACCEPTED) {
+                try {
+                    if (mNumRetries < DEFAULT_NUM_RETRIES) {
+                        mNumRetries++;
+                        // First try with retry-after value, then use exponential back-off
+                        String value = response.getHttpURLConnection().getHeaderField("Retry-After");
+                        if (!SdkUtils.isBlank(value)) {
+                            int retryAfterSeconds = Integer.parseInt(value);
+                            retryAfterSeconds = retryAfterSeconds > 0 ? retryAfterSeconds : 1;
+                            mRetryAfterMillis = retryAfterSeconds * 1000;
+                        }
+                    } else if (mRetryAfterMillis < DEFAULT_MAX_WAIT_MILLIS * 1000) {
+                        mRetryAfterMillis *= 2;
+                    } else {
+                        // Give up after the maximum retry time is exceeded.
+                        throw new BoxException.MaxAttemptsExceeded("Max wait time exceeded.", mNumRetries);
+                    }
+                    Thread.sleep(mRetryAfterMillis);
+                    return (BoxDownload) mRequest.send();
```

Comment:
Let's talk about it tomorrow. I just want to make sure the original
response object is getting populated correctly even with the recursion.
On Jun 10, 2015 8:09 PM, "brianytang" <notifications@github.com> wrote:

> In
> box-content-sdk/src/main/java/com/box/androidsdk/content/requests/BoxRequestDownload.java
> <https://github.com/box/box-android-sdk/pull/16#discussion_r32186804>:
>
> > +                        mNumRetries++;
> > +                        // First try with retry-after value, then use exponential back-off
> > +                        String value = response.getHttpURLConnection().getHeaderField("Retry-After");
> > +                        if (!SdkUtils.isBlank(value)) {
> > +                            int retryAfterSeconds = Integer.parseInt(value);
> > +                            retryAfterSeconds = retryAfterSeconds > 0 ? retryAfterSeconds : 1;
> > +                            mRetryAfterMillis = retryAfterSeconds * 1000;
> > +                        }
> > +                    } else if (mRetryAfterMillis < DEFAULT_MAX_WAIT_MILLIS * 1000) {
> > +                        mRetryAfterMillis *= 2;
> > +                    } else {
> > +                        // Give up after the maximum retry time is exceeded.
> > +                        throw new BoxException.MaxAttemptsExceeded("Max wait time exceeded.", mNumRetries);
> > +                    }
> > +                    Thread.sleep(mRetryAfterMillis);
> > +                    return (BoxDownload) mRequest.send();
>
> So BoxRequest#send() should execute synchronously and would only create a
> new future task if we called toTask() on it.
>
> I've tested this with the Browse sdk and the thumbnail fetching logic
> which uses a different executor and the results are being returned as
> expected. Lemme know if I'm missing something
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/box/box-android-sdk/pull/16/files#r32186804>.
>